src/sensor_interface.c

```diff
@@ -0,0 +1,241 @@
+ */
+void update_sensor_data(const struct device *dev) {
+	// Variables to store sensor data.
+	struct sensor_value temp_value, press_value;
+	// struct sensor_value humidity_value;
+
+	// Check if the device is ready.
+	if (dev != NULL) {
+		// LOG
+		LOG_DBG("Fetching sensor data.");
+		// Read the sensor data.
+		sensor_sample_fetch(dev);
+		sensor_channel_get(dev, SENSOR_CHAN_AMBIENT_TEMP, &temp_value);
+		sensor_channel_get(dev, SENSOR_CHAN_PRESS, &press_value);
+		// sensor_channel_get(dev, SENSOR_CHAN_HUMIDITY, &humidity_value);
+
+		// Print the sensor data.
+		LOG_DBG("Temperature: %.2f C, Pressure: %.2f Pa", 
+			sensor_value_to_double(&temp_value),
+			sensor_value_to_double(&press_value));
+
+	} else {
+		// If the device is not ready, or not found, sample the sensor 
+		// with dummy simulated values.
+		// printk("INFO: Simulating sensor data.");
+
+		// Sample the sensor with dummy values.
+		temp_value.val1 = sim_temperature;
+		temp_value.val2 = sim_temperature * 2 * 1000;
+		press_value.val1 = sim_pressure;
+		press_value.val2 = sim_pressure * 2 * 1000;
+		// humidity_value.val1 = sim_humidity;
+		// humidity_value.val2 = sim_humidity * 2 * 1000;
+
+		// Decrease the dummy values.
+		sim_temperature -= 1;
+		sim_pressure -= 3;
+		// sim_humidity -= 2;
+
+		// If the dummy values are below 0, reset them.
+		if (sim_temperature < 0)
+			sim_temperature = DEFAULT_TEMPERATURE;
+		if (sim_pressure < 0)
+			sim_pressure = DEFAULT_PRESSURE;
+		// if (sim_humidity < 0)
+		//	sim_humidity = DEFAULT_HUMIDITY;
+	}
+
+	// Debug LOG
+	LOG_DBG("Temperature: %.2f C, Pressure: %.2f Pa", 
+		sensor_value_to_double(&temp_value),
+		sensor_value_to_double(&press_value)
+	//	sensor_value_to_double(&humidity_value)
+	);
+
+	// Create a struct to store the sensor readings.
+	struct sensor_value_store_t tx_data = {
+		.temperature = temp_value,
+		.pressure = press_value
+	//	.humidity = humidity_value,
+	};
+
+	size_t datatype_size = sizeof(struct sensor_value_store_t);
+	struct sensor_value_store_t *mem_ptr = k_malloc(datatype_size);
+	__ASSERT_NO_MSG(mem_ptr != 0);
+	memcpy(mem_ptr, &tx_data, datatype_size);
+
+	k_lifo_put(&sensor_value_lifo, mem_ptr);
+}
+
+/* Read the queued sensor values for a given window size, and return the
+ * median values for each sensor as a values_for_ble_t pointer. The returned
+ * pointer must be freed by the caller. This function's output is used to
+ * update the BLE advertisement data.
+ * 
+ * @param window_size The window size to use for the median filter.
+ * @return Pointer to the values_for_ble_t struct, or NULL if the window size
+ *        is not an odd number.
+ */
+struct values_for_ble_t* filter_sensor_value(uint8_t window_size) {
+	// Check if the window size is an odd number.
+	if (window_size % 2 == 0) {
+		LOG_ERR("Window size must be an odd number.");
+		return NULL;
+	}
+
+	// Create a window to store the sensor values.
+	uint8_t temp_window_index = 0;
+	struct sensor_value temp_window[window_size];
```

Comment:
The function parameter defined as a constant in the main code. Have you checked that? Having a function parameter does not mean that it is run-time declared. Did you check if it's building correctly?

Reply:
Yeah, my mistake. It compiles but not because it was a constant-time declared, it is still run-time declared but it uses variable-lenght array feature of the C. I didn't know this feature was exists.

https://en.wikipedia.org/wiki/Variable-length_array